Python Rebuff SDK

mehrinkiani opened the pull request:
This PR adds a first class support for Python SDK for Rebuff.

Please note you will need the following credentials for testing the PR:
`pinecone_environment`
`pinecone_apikey`
`pinecone_index`
`openai_model` 
`openai_apikey`

I will open new issues for creating support for Chroma and for adding tests for Python SDK.

ristomcgehee commented on python-sdk/rebuff/rebuff_pythonsdk.py:
```diff
@@ -0,0 +1,237 @@
+from pydantic import BaseModel
+
+
+class Rebuff_Detection_Response(BaseModel):
```
Since this code isn't parsing user input or doing data validation, I don't see any benefit to inheriting from `BaseModel`. Is there another reason why there's benefit to inheriting from that?

mehrinkiani replied:
I think there was some discussion for keeping one (likely Python) schema and generating ts-schema from it where `BaseModel` might help. Though still a work in progress.

ristomcgehee commented on python-sdk/rebuff/rebuff_pythonsdk.py:
```diff
@@ -0,0 +1,237 @@
+            check_llm (bool, optional): Whether to run the language model check. Defaults to True.
+
+        Returns:
+            Tuple[Union[DetectApiSuccessResponse, ApiFailureResponse], bool]: A tuple containing the detection
```
I believe you copied this line from `rebuff.py` but it's actually incorrect in that file. Here is should be `Rebuff_Detection_Response`.

ristomcgehee commented on python-sdk/rebuff/detect_pi_vectorbase.py:
```diff
@@ -0,0 +1,66 @@
+        vector_score = {
+            "top_score": None,
+            "count_over_max_vector_score": None,
+            "error": error,
```
Since this function is not a web API, I'd recommend just raising an exception instead of including that as an field in the dictionary.

mehrinkiani replied:
Makes sense, thanks. Updated this and other instances (of adding error as a field) to raising exceptions.

ristomcgehee commented on python-sdk/rebuff/detect_pi_vectorbase.py:
```diff
@@ -0,0 +1,66 @@
+        return {"vector_store": vector_store, "error": None}
+
+    except Exception as error:
+        return {"vector_store": None, "error": error}
```
Similarly here, I'd recommend just raising an error.

ristomcgehee reviewed (commented):
Hi Mehrin, I don't think we've interacted yet, but I'm Risto and I've been contributing a little to this project. I think that's great that you're adding a Python SDK! I thought I'd go through and add some comments for where I think the code could be slightly improved. Let me know if you have any questions or clarifications on my review.

mehrinkiani commented:
> Hi Mehrin, I don't think we've interacted yet, but I'm Risto and I've been contributing a little to this project. I think that's great that you're adding a Python SDK! I thought I'd go through and add some comments for where I think the code could be slightly improved. Let me know if you have any questions or clarifications on my review.

Hi Risto, Thank you so much for your review and valuable contributions to the project. I have tried to work on your review. The main changes include:

1. Updated the API call to call Pinecone directly in log_leakage function. 
2. Updated type hints.
3. Changed class name from Rebuff to RebuffPython to avoid possible conflict.

mehrinkiani commented on python-sdk/rebuff/rebuff_pythonsdk.py:
```diff
@@ -0,0 +1,237 @@
+    injection_detected: bool
+
+
+class Rebuff:
```
Thank you for highlighting this. For now, I have updated it to `RebuffPython` to avoid possible conflict.

ristomcgehee replied:
Or `RebuffSdk` would be good to match JS.

mehrinkiani commented on python-sdk/rebuff/rebuff_pythonsdk.py:
```diff
@@ -0,0 +1,237 @@
+        check_heuristic: bool = True,
+        check_vector: bool = True,
+        check_llm: bool = True,
+    ) -> None:
```
Updated it to `Rebuff_Detection_Response`

mehrinkiani commented on python-sdk/rebuff/rebuff_pythonsdk.py:
```diff
@@ -0,0 +1,237 @@
+            "canaryWord": canary_word,
+        }
+        response = requests.post(
+            f"{self.api_url}/api/log", json=data, headers=self._headers
```
Yes, thank you, I have updated it to call Pinecone directly.

mehrinkiani commented on python-sdk/rebuff/rebuff_pythonsdk.py:
```diff
@@ -0,0 +1,237 @@
+        try:
+            import langchain
+
+            if isinstance(prompt, langchain.PromptTemplate):
```
I have updated it to `langchian.prompts.PromptTemplate`

mehrinkiani commented on python-sdk/rebuff/detect_pi_vectorbase.py:
```diff
@@ -0,0 +1,66 @@
+# https://api.python.langchain.com/en/latest/vectorstores/langchain.vectorstores.pinecone.Pinecone.html
+def detect_pi_using_vector_database(
+    input: str, similarity_threshold: float, vector_store: Pinecone
+) -> Union[Dict[str, int], str]:
```
I have updated the type hint to `Dict`, and added in comment the description of dict's keys and values.

ristomcgehee commented on python-sdk/rebuff/rebuff_pythonsdk.py:
```diff
@@ -0,0 +1,228 @@
+            rebuff_heuristic_score = 0
+
+        if check_vector:
+            vector_store = init_pinecone(
```
I'm not sure how much performance would be gained, but you could call `init_pinecone` once and save the `vector_store` in an instance attribute. That's how we do it in the JavaScript SDK.

mehrinkiani replied:
Thank you for the suggestion, I have now added `vector_store` as an instance attribute.

ristomcgehee replied:
I'd recommend putting:
```py
        if self.vector_store is None:
            self.vector_store = init_pinecone(
                self.pinecone_environment,
                self.pinecone_apikey,
                self.pinecone_index,
                self.openai_apikey,
            )
```
into a helper function.

ristomcgehee commented on python-sdk/rebuff/rebuff_pythonsdk.py:
```diff
@@ -0,0 +1,228 @@
+            )
+
+            rebuff_vector_score = 0
+            similarity_threshold = 0.3
```
Shouldn't this be:
```suggestion
            similarity_threshold = max_vector_score
```

mehrinkiani replied:
Yes, thank you, I have updated it!

ristomcgehee commented on python-sdk/rebuff/rebuff_pythonsdk.py:
```diff
@@ -0,0 +1,228 @@
+        vector_store._text_key = "input"
+        vector_store.add_texts(
+            [user_input],
+            metadatas=[{"completion": completion, "canary_word": canary_word}],
```
I think we should make sure the metadata has the same keys whether it's coming from JS or Python (in JS, the key is `"canaryWord"` instead of `"canary_word"`). I'd actually recommend changing the JS to match what you have in Python here since snake case is more common for names of data fields than camel case.

mehrinkiani replied:
Yes, I think we would want the metadata keys to be same. Though I have not updated the JS code yet since I am not sure how [metadata keys](https://github.com/protectai/rebuff/blob/main/javascript-sdk/src/sdk.ts#L152C50-L152C60) are being setup in JS.

ristomcgehee replied:
In TypeScript, when you initialize a `Record` (same thing as a `dict` in Python) using this syntax:
```ts
const metadata = { completion, canaryWord };
```
the variable name becomes the key and the variable value becomes the value. It's the same as doing:
```ts
const metadata = { "completion": completion, "canaryWord": canaryWord };
```
So, you could change the line you linked to:
```ts
        this.logLeakage(userInput, { completion, "canary_word": canaryWord });
```
If you like, I can do this change in a separate PR.

mehrinkiani replied:
Thank you for the detailed explanation. I have made the change as you suggested, though would appreciate if you can review it again please.

ristomcgehee replied:
Yep, that change looks good. :+1:

ristomcgehee commented on python-sdk/rebuff/rebuff_pythonsdk.py:
```diff
@@ -0,0 +1,228 @@
+        return None
+
+
+def encode_string(message: str) -> str:
```
You could delete this since it's not being called.

ristomcgehee commented on python-sdk/rebuff/rebuff_pythonsdk.py:
```diff
@@ -0,0 +1,228 @@
+import langchain
+
+
+class Rebuff_Detection_Response(BaseModel):
```
Naming convention is to not use underscores in class names: 
```suggestion
class RebuffDetectionResponse(BaseModel):
```

ristomcgehee commented on python-sdk/rebuff/rebuff_pythonsdk.py:
```diff
@@ -0,0 +1,228 @@
+    injection_detected: bool
+
+
+class RebuffPython:
```
I believe you'll need to include this class and `RebuffDetectionResponse` in `python-sdk/rebuff/__init__.py` in order for those classes to be imported from outside this module.

ristomcgehee reviewed (commented):
Thanks for addressing my comments. I have a few more... :grin:

mehrinkiani commented on python-sdk/rebuff/rebuff_pythonsdk.py:
```diff
@@ -0,0 +1,228 @@
+                return prompt, canary_word
+
+        except Exception as error:
+            raise Exception(error)
```
Makes sense, thank you. For this code snippet, I am now raising a `TypeError` if `prompt` is neither `str` nor `langchain.prompts.PromptTemplate`. Though I have updated the other instances of try-except as per your suggestion.

mehrinkiani commented on python-sdk/rebuff/rebuff_pythonsdk.py:
```diff
@@ -0,0 +1,228 @@
+
+            rebuff_vector_score = 0
+            similarity_threshold = 0.3
+            vector_store._text_key = "input"
```
The initialization of Pinecone client requires a `_text_key` field.  I am not sure what is the `_text_key` exactly for though I have added the [reference](https://github.com/langchain-ai/langchain/blob/a6ebffb69504576a805f3b9f09732ad344751b89/langchain/vectorstores/pinecone.py#L57) as a comment in code as well.

ristomcgehee replied:
Ok, that link was helpful. `_text_key` is being use as the key for a metadata field that always gets added to a vector when it's inserted. If you don't specify it, it defaults to `"text"` as the metadata key. In the JS SDK, we're not specifying any value, so it's using the default value of `"text"`. Since we want Python and JS to match, I think you should remove this line.

mehrinkiani replied:
Thank you for elaborating on it, it makes more sense now. I think `_text_key` is also being used when [searching for similar documents](https://github.com/langchain-ai/langchain/blob/a6ebffb69504576a805f3b9f09732ad344751b89/langchain/vectorstores/pinecone.py#L126). Hence, if I remove the said line, I get the error: `Found document with no 'text' key. Skipping.`

@seanpmorgan can probably shed more light on it but I think the previous/earlier entries in Pinecone have been made against `self._text_key = 'input'`

ristomcgehee replied:
Ah, I see. In the particular Pinecone database you're testing against, the entries have a metadata field of `"input"` but not `"text"`, so the results all get skipped unless you set `self._text_key = 'input'`.

I understand why Langchain has this behavior (when querying a vector database, you almost always want the text content of the vector), but in our use case, we're not actually doing anything with the text content of the vector, just the score. I think the correct way forward at this point is to submit a PR to Langchain to have the option to not skip results when the metadata doesn't contain the text key (I'll submit that PR). I think you can still move forward with this PR in the meantime by instead of doing `self._text_key = 'input'`, do:
```py
vector_store = Pinecone.from_existing_index(index, openai_embeddings, text_key="input")
```
in `init_pinecone`. Whenever possible, it's better to set attributes via constructor or function arguments instead of directly on the instance (in Python, it's a convention that attributes that start with an underscore should not be accessed outside of that class).

I looked at the [JS code](https://github.com/langchain-ai/langchainjs/blob/8154c7829fc930a5b4060e830d06cbdd3e9484c4/libs/langchain-community/src/vectorstores/pinecone.ts#L230), and Langchain does not skip vectors when `this.textKey` is not found in the metadata, so we don't need to make any JS changes in regards to this issue.

ristomcgehee replied:
Langchain PR: https://github.com/langchain-ai/langchain/pull/15663

mehrinkiani replied:
Thank you for elaborating on it, and for submitting the Langchain PR. I have updated `init_pinecone` as you suggested.

mehrinkiani commented on python-sdk/rebuff/detect_pi_openai.py:
```diff
@@ -0,0 +1,83 @@
+    """
+    openai.api_key = api_key
+
+    completion = openai.ChatCompletion.create(
```
Thank you for highlighting this. I have updated the code for updated `openai` API. Also, I noticed that the class [`OpenAIEmbeddings` is deprecated](https://api.python.langchain.com/en/latest/embeddings/langchain_community.embeddings.openai.OpenAIEmbeddings.html#) in v0.1.0 and might get removed in v0.2.0 so also updated code with the alternative they have suggested: `langchain_openai.OpenAIEmbeddings` . I have also added `langchain_openai` in `setup.py` under `install_requires`

ristomcgehee replied:
That's great!

ristomcgehee commented:
This PR looks good from my perspective. Thanks for all your work on it, Mehrin!

seanpmorgan reviewed (approved):
LGTM, Thanks @mehrinkiani and @ristomcgehee for the review!
